File: python/tennis-ball-detector/detector.py
import cv2
import sys
import os

directory = sys.argv[1]
classifier = sys.argv[2]

classifier = cv2.CascadeClassifier(classifier)
images_to_check = os.listdir(directory)

if(not os.path.exists('results')):
    os.makedirs('results')

for image in images_to_check:
    img = cv2.imread(f"{directory}/{image}")
    img_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    found = classifier.detectMultiScale(img_gray, minSize =(40, 40))
    if(len(found)!=0):
        with open(f"results/{image[:-4]}_balls.txt", 'w') as f:
            for i, (x, y, width, height) in enumerate(found):
                cv2.rectangle(img, (x, y), (x + height, y + width), (0, 255, 0), 5)
                f.write(f"ball_no_{i}: coords ({x}, {y}), size {height}x{width}\n")
    cv2.imwrite(f"results/{image}", img)
    

File: tennis-ball-detector/detector.cpp
#include <opencv2/imgcodecs.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/objdetect.hpp>
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <filesystem>
#include <stringstream>

using namespace cv;
using namespace std;

int main()
{
    string classifier = "data/cascade.xml";

    CascadeClassifier ballCascade;
    ballCascade.load(classifier);

    for (int num = 14; num <= 30; num++)
    {
        stringstream ss;
        ss << "image_" << num << ".png";

        string path = ss.str;
        Mat img, imgGray;
        img = imread(path);
        cvtColor(img, imgGray, COLOR_BGR2GRAY);

        vector<Rect> balls;
        ballCascade.detectMultiScale(imgGray, faces, 2, 2);

        for (int i = 0; i < faces.size(); i++)
        {
            rectangle(img, faces[i].tl(), faces[i].br(), Scalar(255, 0, 255), 3);
        }

        imshow("found balls", img);
        waitKey(0);
    }
}